File: spark-wordcount-java/src/main/java/org/cooze/hadoop/spark/wordcount/java/WordCount.java
package org.cooze.hadoop.spark.wordcount.java;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

import java.util.Arrays;
import java.util.List;
import java.util.Iterator;

/**
 * @author cooze
 * @version 2.0.0 创建于 2017/10/13
 */
public class WordCount {


    public static void main(String[] args) {


        String appName = "WordCount";
        String master = "local";

        SparkConf conf = new SparkConf().setAppName(appName).setMaster(master);
        JavaSparkContext sc = new JavaSparkContext(conf);

        //待处理文件路径
        String files = args[0];
        JavaRDD<String> rdd = sc.textFile(files);

        //映射行
        JavaRDD<String> words = rdd.flatMap((new FlatMapFunction<String, String>() {
            public Iterator<String> call(String s) throws Exception {
                return Arrays.asList(s.split(" ")).iterator();
            }
        }));


        //映射单个单词
        JavaPairRDD<String, Integer> ones = words.mapToPair(new PairFunction<String, String, Integer>() {
            public Tuple2<String, Integer> call(String s) {
                return new Tuple2<String, Integer>(s, 1);
            }
        });

        //单词汇总
        JavaPairRDD<String, Integer> counts = ones.reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer i1, Integer i2) {
                return i1 + i2;
            }
        });

        //结果输出
        List<Tuple2<String, Integer>> output = counts.collect();

        for (Tuple2<?, ?> tuple : output) {
            System.out.println(tuple._1() + ": " + tuple._2());
        }

        sc.stop();

    }
}

File: spark-wordcount-scala/src/main/scala/org/cooze/hadoop/spark/wordcount/scala/WordCount.scala
package org.cooze.hadoop.spark.wordcount.scala

import org.apache.spark.{SparkConf, SparkContext}

/**
  * Created by cooze on 2017/10/13.
  */
object WordCount {
  def main(args: Array[String]): Unit = {
    println("Hello WordCount programmer.")

    val appName = "WordCount"
    val master = "local"

    val conf = new SparkConf().setAppName(appName).setMaster(master)
    val context = new SparkContext(conf)

    //传入的第一个参数就是待计算文件路径
    val filePath = args(0)

    val rdd = context.textFile(filePath)

    //spark rrd
    rdd.flatMap(_.split(" ")).map((_, 1)).reduceByKey(_ + _).collect().foreach(println)

    context.stop()

  }

}
